Example cannot be imported in Windows if python >= 3.8

Some while ago, I marked the test as allowed to fail, but it would be wise to open an issue to track the problem.

The SWIG example compiles fine in all Window versions, generating the following files:

```
adding 'mymath/_bindings.pyd'
adding 'mymath/bindings.py'
adding 'mymath/libbindings.dll.a'
adding 'mymath/include/mymath.h'
adding 'mymath/lib/libmymath.a'
adding 'mymath-0.0.0.dist-info/METADATA'
adding 'mymath-0.0.0.dist-info/WHEEL'
adding 'mymath-0.0.0.dist-info/top_level.txt'
adding 'mymath-0.0.0.dist-info/RECORD'
```

However, when it is imported in the tests, it fails with the following error (more details in any CI job):

```
=================================== ERRORS ====================================
____________________ ERROR collecting tests/test_mymath.py ____________________
ImportError while importing test module 'D:\a\cmake-build-extension\cmake-build-extension\examples\swig\tests\test_mymath.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
c:\hostedtoolcache\windows\python\3.9.5\x64\lib\importlib\__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
tests\test_mymath.py:3: in <module>
    import mymath.bindings
c:\hostedtoolcache\windows\python\3.9.5\x64\lib\site-packages\mymath\bindings.py:18: in <module>
    from . import _bindings
E   ImportError: DLL load failed while importing _bindings: The specified module could not be found.
=========================== short test summary info ===========================
ERROR tests/test_mymath.py
!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!
============================== 1 error in 0.37s ===============================
```

I have no idea what could cause this error. I marked as allowed to fail also the new Python 3.9 job.

---

I also caught the following in the Python 3.7 job:

```
C:\Users\RUNNER~1\AppData\Local\Temp\pip-build-env-t1xho44w\overlay\Lib\site-packages\wheel\bdist_wheel.py:82: RuntimeWarning: Config variable 'Py_DEBUG' is unset, Python ABI tag may be incorrect
    warn=(impl == 'cp')):
  C:\Users\RUNNER~1\AppData\Local\Temp\pip-build-env-t1xho44w\overlay\Lib\site-packages\wheel\bdist_wheel.py:87: RuntimeWarning: Config variable 'WITH_PYMALLOC' is unset, Python ABI tag may be incorrect
    sys.version_info < (3, 8))) \
```

The second warning disappears if Python >= 3.8.

--- a/.github/workflows/python.yml
+++ b/.github/workflows/python.yml
@@ -23,14 +23,6 @@ jobs:
           - ubuntu-20.04
           - macos-latest
           - windows-latest
-        include:
-          # https://github.com/diegoferigo/cmake-build-extension/issues/8
-          - os: windows-latest
-            python-version: 3.8
-            experimental: true
-          - os: windows-latest
-            python-version: 3.9
-            experimental: true
 
     steps:
 

--- a/examples/swig/setup.cfg
+++ b/examples/swig/setup.cfg
@@ -20,7 +20,9 @@ packages = find:
 package_dir =
     =src
 python_requires = >=3.6
-install_requires = numpy
+install_requires =
+    numpy
+    cmake-build-extension
 
 [options.packages.find]
 where = src

--- a/examples/swig/setup.py
+++ b/examples/swig/setup.py
@@ -1,3 +1,4 @@
+import inspect
 import sys
 from pathlib import Path
 
@@ -7,11 +8,21 @@ from setuptools import setup
 with open(Path(__file__).parent.absolute() / "README.md", encoding="utf-8") as f:
     long_description = f.read()
 
+init_py = inspect.cleandoc(
+    """
+    from cmake_build_extension import build_extension_env
+    
+    with build_extension_env():
+        from . import bindings
+    """
+)
+
 setup(
     ext_modules=[
         CMakeExtension(
             name="mymath",
             install_prefix="mymath",
+            write_top_level_init=init_py,
             source_dir=str(Path(__file__).parent.absolute()),
             cmake_configure_options=[
                 f"-DPython3_ROOT_DIR={Path(sys.prefix)}",

--- a/src/cmake_build_extension/__init__.py
+++ b/src/cmake_build_extension/__init__.py
@@ -1,3 +1,48 @@
+import os
+from contextlib import contextmanager
+from pathlib import Path
+
 from . import build_ext_option
 from .build_extension import BuildExtension
 from .cmake_extension import CMakeExtension
+
+
+@contextmanager
+def build_extension_env():
+    """
+    Creates a context in which build extensions can be imported.
+
+    It fixes a change of behaviour of Python >= 3.8 in Windows:
+    https://docs.python.org/3/whatsnew/3.8.html#bpo-36085-whatsnew
+
+    Other related resources:
+
+    - https://stackoverflow.com/a/23805306
+    - https://www.mail-archive.com/dev@subversion.apache.org/msg40414.html
+
+    Example:
+
+    .. code-block:: python
+
+        from cmake_build_extension import build_extension_env
+
+        with build_extension_env():
+            from . import bindings
+    """
+
+    cookies = []
+
+    # Windows and Python >= 3.8
+    if hasattr(os, "add_dll_directory"):
+
+        for path in os.environ.get("PATH", "").split(os.pathsep):
+
+            if path and Path(path).is_absolute() and Path(path).is_dir():
+                cookies.append(os.add_dll_directory(path))
+
+    try:
+        yield
+
+    finally:
+        for cookie in cookies:
+            cookie.close()

